keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,510 @@
+defaulting purposes.
+
+We propose the creation of a new CRD flag: `preserveNullValues` which
+will default to `true`. This flag will be used as a ratcheting mechanism
+as we migrate to this new semantic throughout all Kubernetes types
+(which would be how all types eventually work). When false, the
+apiserver will use new semantics and remove all null values from CR
+objects. When true, the existing and deprecated semantics will be kept.
+That new flag is following the same logic as "preserveUnknownFields" and
+the new operation will happen using the same mechanisms right after
+pruning.
+
+Note that we still consider `{}` (and 0, 0.0, "", false, []) to be
+different from `null`, so `{empty: {}}` is different from `{}` or from
+`{empty: null}` (but the latter two are considered equal).
+
+As described by @sttts on many occasions, the immutability is blocked
+because we don't know how to distinguish between null and
+unspecified. We believe that this mechanism will be able to unblock it.
+
+### Built-in Defaulting
+
+Because built-in types are serialized and deserialized as Go structures,
+there are semantics that are imposed on us. The two main ones are:
+1. We can't differentiate between null, unspecified and zero-value,
+2. non-pointer structs are always present (and hence have to be
+recursively defaulted), they also can't be omitEmpty'd.
+
+The first point is solved by the removed nulls described in the previous
+section (and invalid zero-values can't be present in objects), and the
+second point can be solved by forcing non-pointer structs to have a `{}`
+default. To address that, we're proposing that the `default` marker is
+forbidden on non-pointer structs (pointer structs can behave as
+expected) and the default field of the openapi is automatically set to
+`{}`. This only applies to OpenAPI generated from built-in types or from
+kubebuilder. CRDs can still have any defaulting they have for
+structures, but we strongly suggest that OpenAPI generated from Go types
+should follow that semantics.
+
+For example:
+```golang
+Type Root struct {
+  // Defaults on non-pointer structs are FORBIDDEN:
+  // +default={name: entry, number: 12}
+  Entry SubLevel
+}
+
+Type SubLevel struct {
+  Name string
+  Number int
+}
+```
+
+Computed defaults (through defaulting functions) will not be supported
+as a declarative marker and will continue being supported through the
+existing defaulting functions.
+
+### Risks and Mitigations
+
+The main risk is that we could end-up with different defaults for
+existing types which would be extremely confusing for our users.
+
+## Design Details
+
+We are proposing to modify the defaulting generator to also generate new
+defaulting functions, which will use the marker to generate the default
+and apply them to the fields directly if the fields are set to their
+zero-value. These new defaulting functions will be called right before
+the existing defaulting functions.
```

Comment:
That's a little unfortunate that it's two phases rather than one phase but humans can edit/override the generated functions if desired.